src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/InvokerUtil.cs

```diff
@@ -150,8 +168,74 @@ internal static InvokeDelegate GenerateInvokeDelegate(MethodInfo method, out int
 
                     return result;
                 };
+            }
+
+            private static InvokeDelegate GenerateInvokeDelegateInternalWithExpressions(int paramCount, bool returnsValue, int[] inputPos, int[] outputPos, MethodInfo method)
+            {
+                var targetParam = Expression.Parameter(typeof(object), "target");
+                var inputsParam = Expression.Parameter(typeof(object[]), "inputs");
+                var outputsParam = Expression.Parameter(typeof(object[]), "outputs");
+                var paramsLocal = Expression.Variable(typeof(object[]), "paramsLocal");
+                var result = Expression.Variable(typeof(object), "result");
+
+                var methodParam = Expression.Constant(method);
+
+                var assignParamsLocal = Expression.Assign(paramsLocal, Expression.Condition(
+                    Expression.GreaterThan(Expression.Constant(paramCount), Expression.Constant(0)),
+                    Expression.NewArrayBounds(typeof(object), Expression.Constant(paramCount)),
+                    Expression.Constant(null, typeof(object[]))));
+
+                var assignParamsLocalValues = inputPos.Length == 0
+                    ? Expression.Block(Expression.Empty())
+                    : Expression.Block(
+                        Enumerable.Range(0, inputPos.Length).Select(i =>
+                            Expression.Assign(
+                                Expression.ArrayAccess(paramsLocal, Expression.Constant(inputPos[i])),
+                                Expression.ArrayIndex(inputsParam, Expression.Constant(i)))));
+
+                var invokeMethod = Expression.Block(
+                    Expression.Assign(result, Expression.Condition(
+                        Expression.Equal(Expression.Constant(returnsValue), Expression.Constant(true)),
+                        Expression.Call(methodParam, nameof(MethodInfo.Invoke), null, targetParam, paramsLocal),
+                        Expression.Block(
+                            Expression.Call(methodParam, nameof(MethodInfo.Invoke), null, targetParam, paramsLocal),
+                            Expression.Constant(null, typeof(object))))),
+                    Expression.Empty());
+
+                var catchParameterExpr = Expression.Parameter(typeof(TargetInvocationException), "tie");
+                var throwCapturedExceptionDispatchInfoExpr = Expression.Call(Expression.Call(typeof(ExceptionDispatchInfo), nameof(ExceptionDispatchInfo.Capture), null,Expression.Property(catchParameterExpr, nameof(Exception.InnerException))), nameof(ExceptionDispatchInfo.Throw), null);
+
+                var tryCatch = Expression.TryCatch(invokeMethod,
```

Comment:
The try/catch should now disappear. It's only needed because calling `MethodInfo.Invoke` wraps the exception that's throw in in a `TargetInvocationException` and we need to unwrap it. If you use `Expression.Call` directly on the `methodInfo` without calling the `Invoke` method, it will directly throw the correct exception and we don't need to unwrap it.

Reply:
great!, done!